### PR TITLE
Refresh token working!!

### PR DESCRIPTION
**Pull request by Lenasim**
Finally it's working.........!!!!!!

## what you need to check to test: 
1) when login : you'll see localStorage => stud-connect@token/username/etc % cookie => refresh token
image
image

2) when logout : you'll see all localStorage and cookie is cleared
image
image

3) Refresh token is automatically executed via 'apollo link'. 
You can see the function is executed when fetch request is called. 
So need to click 'forum' to see refresh works where apollo query is called.
(You can also see if '/refresh_token' route is called in inspection => Network tap)
Little tip to test => change env value JWT_LIFE_TIME to 5s (it will execute the function in about 5seconds)

## !MIND! using local server (branch: auth) to test as PR in the back is not yet accepted.
  So you'll see  index.js, uri's are commented or changed like below. 
  it can be fixed in clean code when PR on the back side is deployed^^
--line 13
`  // uri: process.env.REACT_APP_GRAPHQL_URI || 'http://localhost:4000/graphql',
  uri: 'http://localhost:4000/graphql',`
--line 39 
`fetch('http://localhost:4000/refresh_token' || 'https://stud-connect.herokuapp.com/refresh_token',`

## Big changes concerning CONFLICT & SHOCK
- apollo client server is moved to 'index.js' from 'app.js' 
- local storage variable name is changed (add 'stud-connect@')
- All pages written in sidebar (By Clem) are added to router and pages as well (with dummy text)
It can be changed as long as you want!! (Especially Nico who are working on the sidebar) 
Home (need to click button GO TO HOME and it will redrect to login page)
image
Dummy page example
image
image

Hope it works!

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -1,43 +1,61 @@
...
+      <RouteWrapper exact path='/topics' history component={ForumContainer} layout={Authenticated} text='Forum' />
+      <RouteWrapper path='/topics/:id' component={TopicContainer} protectedRoute layout={Authenticated} />
+    </Switch>
   );
 }
 
-function RouteWrapper({ component: Component, layout: Layout, text, ...rest }) {
+function RouteWrapper(props) {
+  const { component: Component, layout: Layout, text, protectedRoute, ...rest } = props;
+  const isAuth = !!localStorage.getItem('stud-connect@userID');
```
Comment (coneeone) [inline src/App.js:47]: Don't know the _!!_ operator and can't find what it stands for.
Would like to discuss it before merging

File: src/index.js [hunk 2]
```diff
@@ -1,15 +1,88 @@
...
+        return false;
+      }
+      return true;
+    } catch {
+      return false;
+    }
+  },
+  fetchAccessToken: () => {
+    const token = localStorage.getItem('stud-connect@token') || null;
+    return fetch('http://localhost:4000/refresh_token' || 'https://stud-connect.herokuapp.com/refresh_token', {
```
Comment (coneeone) [inline src/index.js:39]: Both strings will always be true so _'http://localhost:4000/refresh_token'_ will always be called
Might use something like : 
```javascript
fetch(process.env === 'dev' ? localhost:4000/refresh_token : process.env.refresh_token_uri)
```

Comment (coneeone) [reply src/index.js:39]: (line 39)

File: src/index.js [hunk 3]
```diff
@@ -1,15 +1,88 @@
...
+      method: 'GET',
+      credentials: 'include',
+      headers: {
+        Authorization: `bearer ${token}` || null,
```
Comment (coneeone) [inline src/index.js:43]: the string _bearer ${token}_ will always be true and will always be returned
Might instead use : 
```javascript
Authorization: token ? `bearer ${token}` : null,
```

Comment (coneeone) [reply src/index.js:43]: (line 43)

File: src/index.js [hunk 4]
```diff
@@ -1,15 +1,88 @@
...
+});
+
+const authMiddleware = new ApolloLink((operation, forward) => {
+  const token = localStorage.getItem('stud-connect@token') || null;
+
+  // add the authorization to the headers
+  if (token) {
+    operation.setContext({
+      headers: {
+        Authorization: `bearer ${token}` || null,
```
Comment (coneeone) [inline src/index.js:65]: the string bearer ${token} will always be true and will always be returned
See comment of the line 43

**Review by coneeone — COMMENTED** [review]
# Very good work 🙏

* when login : you'll see localStorage => stud-connect@token/username/etc % cookie => refresh token ✅
* when logout : you'll see all localStorage and cookie is cleared ✅
* Refresh token is automatically executed via 'apollo link'. ✅

### bug report 🐜
Seems like there is a side effect with _ForumContainer_ when it is unmounted by the logout.
Not critical, can be fixed after merging
```
Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
    at ForumContainer (http://localhost:3000/static/js/main.chunk.js:2758:3)
    at div
    at div
    at div
    at Authenticated (http://localhost:3000/static/js/main.chunk.js:5827:3)
```

File: src/App.js (hunk 1 above)
Comment (coneeone) [reply src/App.js:47]: (line 47)

File: src/index.js (hunk 4 above)
Comment (coneeone) [reply src/index.js:65]: (line 65)

Comment (Lenasim) [reply src/index.js]: correction done

File: src/index.js (hunk 3 above)
Comment (Lenasim) [reply src/index.js]: correction done

File: src/index.js (hunk 2 above)
Comment (Lenasim) [reply src/index.js]: correction done

File: src/App.js (hunk 1 above)
Comment (Lenasim) [reply src/App.js:47]: to return in boolean (true when value exist)

**Pull request merged**